frame/contracts/src/wasm/mod.rs

```diff
@@ -229,68 +226,122 @@ where
 
 		// Instantiate the instance from the instrumented module code and invoke the contract
 		// entrypoint.
-		let input_data_cloned = input_data.clone();
-
-		// TODO(hussein-aitlahcen): avoid panicking, even if the contract has been previously
-		// checked for the export.
-		let make_runtime = |not_started_instance: &NotStartedModuleRef| match not_started_instance
-			.not_started_instance()
-			.export_by_name("memory")
-			.expect("impossible")
+		let mut runtime = Runtime::new(ext, &code, &imports, input_data.clone())?;
+
+		pub fn to_string<T>(data: &T) -> Result<alloc::string::String, DispatchError>
+		where
+			T: serde::ser::Serialize + ?Sized,
 		{
-			ExternVal::Memory(memory) => {
-				log::debug!(target: "runtime::contracts", "Set internal memory");
-				Runtime::new(ext, input_data_cloned, Memory::new(memory))
+			serde_json::to_string(data).map_err(|_| DispatchError::Other("couldn't serialize"))
+		}
+
+		log::debug!(target: "runtime::contracts", "Executing function {:?}", function);
+
+		match function {
+			ExecuteFunction::Instantiate => {
+				// TODO(hussein-aitlahcen): example of env setup + instantiation
+				// this should probably be setup by the caller as well.
+
+				let env = Env {
+					block: BlockInfo {
+						height: 0,
+						time: Timestamp(
+							to_string(&0u128)
+								.map_err(|_| DispatchError::Other("marshall failed"))?,
+						),
+						chain_id: Default::default(),
+					},
+					transaction: Some(TransactionInfo { index: 0 }),
+					contract: ContractInfo {
+						address: Addr::unchecked(T::AccountIdToFromString::convert(
+							runtime.ext().address().clone(),
+						)),
+					},
+				};
+
+				let info = MessageInfo {
+					sender: Addr::unchecked(T::AccountIdToFromString::convert(
+						runtime.ext().caller().clone(),
+					)),
+					funds: vec![],
```

Comment:
Should we pass the funds as an argument from the function here?

Reply:
The funds are only 1 token in the contracts pallet while it's a vec in cosmwasm, we can pass `vec![("PICA", transferred_amount)]` already :+1: .
But I think when rewritting, we'll make it multi-currency and correctly populate the list of coins with all the transferred coins.